### ISSUE TITLE
Potential huge loop in PermutationGroup

### ISSUE THREAD

**Issue by lafita** [biojava/biojava#321]
In determining the **RotationGroup** of an assembly permutations of all the subunits are generated by combining existing valid permutations. The maximum number of permutations is the factorial of the number of subunits, but this is not taken into account in the recursive method *completeGroup()* inside **PermutationGroup**. This means that permutations continue to be combined even if the number of existing permutations is the maximum (so no new ones will be found).

For an 8 subunit example, the maximum number of permutations is 40320 (a feasible number), but if all of them have to be combined to each other (an avoidable situation) the number of iterations rises to nearly 2 billion. 

However, this small fix (checking if the maximum number of permutations has been reached) does not solve cases with higher number of subunits (with 13 subunits, 6 billion permutations possible). Other solutions might be a maximum number of permutations or a timeout threshold, like for the local symmetry calculation.

Although these are rare cases, they correspond to the worst case scenario for the method. An example is the domain d1tnsa_. To reproduce the results run the internal symmetry analysis and try to calculate the PG:

```java
String name = "d1tnsa_";

AtomCache cache = new AtomCache();
Structure s = cache.getStructure(name);
Atom[] atoms = StructureTools.getRepresentativeAtomArray(s);

CeSymm ceSymm = new CeSymm();
CESymmParameters params = ceSymm.getParameters();
params.setOptimization(true);
params.setMultipleAxes(false);

MultipleAlignment result = ceSymm.align(atoms, params);

//This line does not terminate (takes very long)
StructureTools.getQuaternarySymmetry(result);
```

**Comment by sbliven** [comment]
@pwrose Do you remember the reasoning behind `PermutationGroup.completeGroup()`, which tries combining all N^2 permutations iteratively? This seems like a perfect place for some clever optimization. Why not just use the standard algorithm to generate all possible permutations of n elements?

**Comment by pwrose** [comment]
In general, when a structure has symmetry, only a few symmetry operations
need to be identified. All others are combinations of existing symmetry
operations (hence the word point group). For the example of icosahedral
symmetry, I think a 2, 3, and 5 fold axis are sufficient to generate all 60
symmetry operations by combining the existing operations.

CompleteGroup usually finishes after a few iterations. I'm not aware of
cases where it would do many permutations. I currently don't have time to
look at your specific case, but I haven't noticed this behavior running it
through all structures in the PDB. Perhaps you encountered some edge case?
Does the structure you pass in have symmetry or partial symmetry? Does it
ever finish? What symmetry does it report?

On Fri, Sep 11, 2015 at 1:37 AM, Spencer Bliven <notifications@github.com>
wrote:

> @pwrose <https://github.com/pwrose> Do you remember the reasoning behind
> PermutationGroup.completeGroup(), which tries combining all N^2
> permutations iteratively? This seems like a perfect place for some clever
> optimization. Why not just use the standard algorithm to generate all
> possible permutations of n elements?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biojava/biojava/issues/321#issuecomment-139485991>.
>



-- 
Peter Rose, Ph.D.
Site Head, RCSB Protein Data Bank West (http://www.rcsb.org)
San Diego Supercomputer Center (http://bioinformatics.sdsc.edu)
University of California, San Diego
+1-858-822-5497

**Comment by sbliven** [comment]
I created a minimal example to profile the function, based on the permutations detected in @lafita's example:

```java
	public static void profileCompleteGroup() {
		int runs=10;
		final int f8 = factorial(8);
		long itime = System.currentTimeMillis();
		for(int i=0;i<runs;i++) {
			PermutationGroup g = new PermutationGroup();
			g.addPermutation(Arrays.asList(0,1,2,3,4,5,6,7));
			g.addPermutation(Arrays.asList(6,5,4,3,2,1,0,7));
			g.addPermutation(Arrays.asList(5,7,0,1,2,3,4,6));
			g.completeGroup();
			assert(g.getOrder() == f8);
		}
		long etime = System.currentTimeMillis();
		
		System.out.format("Avg Time for %d runs: %.4f s",runs, (etime-itime)/1000./runs );
	}
```

Note that all the quaternary examples in the PDB have low order (<=60), so such a case wouldn't come up. Here we have a particularly poor alignment which leads to all `8!` permutations being generated.

This example didn't finish in the 2 hours I let it run. One inefficiency is in the addPermutation method, which uses the O(N) `List.contains()` method. Replacing this with a `HashSet` to test membership led to a significant speedup to 600 s. Then I replaced the nested loops with a breadth-first search of the graph of permutations, and now the method takes 0.05 s. It's about 50% faster for the typical small cases too.

**Comment by pwrose** [comment]
Nice solution, Spencer!

On Tue, Sep 15, 2015 at 6:07 AM, Spencer Bliven <notifications@github.com>
wrote:

> Closed #321 <https://github.com/biojava/biojava/issues/321> via 209f41d
> <https://github.com/biojava/biojava/commit/209f41d0f880c664bae9371df28c3558b0a93ef2>
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biojava/biojava/issues/321#event-409759690>.
>



-- 
Peter Rose, Ph.D.
Site Head, RCSB Protein Data Bank West (http://www.rcsb.org)
San Diego Supercomputer Center (http://bioinformatics.sdsc.edu)
University of California, San Diego
+1-858-822-5497
